Add support for channel categories

pegasy opened the pull request:
Implemented as its own channel type

pegasy committed:
Add support for channel categories (as its own channel type)

AntiTcb commented on src/Discord.Net.Rest/Entities/Channels/RestGuildChannelCategory.cs:
```diff
@@ -0,0 +1,38 @@
+            base.Update(model);
+        }
+
+        public Task<RestGuildUser> GetUserAsync(ulong id, RequestOptions options = null)
```
Actions that cannot be performed on categories that are being inherited from base classes/interfaces should throw a `NotSupportedException`. For categories, this includes getting users and creating invites.

AntiTcb commented on src/Discord.Net.WebSocket/Entities/Channels/SocketGuildChannel.cs:
```diff
@@ -17,6 +17,9 @@ public class SocketGuildChannel : SocketChannel, IGuildChannel
         public int Position { get; private set; }
+        public ulong? ParentId { get; private set; }
+        public IGuildChannel ParentChannel => ParentId == null ? null : Guild.GetChannel(ParentId.Value);
+        public Task<IGuildChannel> GetParentChannelAsync() => Task.FromResult(ParentChannel);
```
This should be an explicit interface implementation of `IGuildChannel` as it is redundant with the public property `ParentChannel`.

AntiTcb replied:
Alongside, the type should be `IChannelCategory` as actual channels can't be parents/children of other channels, only categories.

AntiTcb commented on src/Discord.Net.Rest/Entities/Guilds/RestGuild.cs:
```diff
@@ -176,6 +176,11 @@ public Task<RestGuildChannel> GetChannelAsync(ulong id, RequestOptions options =
+        public async Task<IReadOnlyCollection<RestGuildChannelCategory>> GetChannelCategoriesAsync(RequestOptions options = null)
+        {
+            var channels = await GuildHelper.GetChannelsAsync(this, Discord, options).ConfigureAwait(false);
+            return channels.Select(x => x as RestGuildChannelCategory).Where(x => x != null).ToImmutableArray();
```
Opt for `ToReadOnlyCollection()` when you have a collection with a `.Count` property, rather using `.ToImmutableArray()`.

pegasy replied:
Only did the same as what was done for the methods right above it (GetVoiceChannelsAsync and GetTextChannelsAsync)

AntiTcb replied:
Ah. Couldn't see them from the review page.

AntiTcb commented on src/Discord.Net.Core/Entities/Channels/IGuildChannelCategory.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Discord
+{
+    public interface IGuildChannelCategory : IGuildChannel
```
Should just be IChannelCategory, RestChannelCategory, etc. Current name is redundant.

pegasy commented:
The field from the api is called parent_id, is this what it should be called throughout discord.net or is that just the technical name for it? Would CategoryId be a better name for it?

AntiTcb commented:
I'd opt for using "Category" over "Parent".

pegasy commented:
ChannelCategory vs simply Category in the context of creating / getting a category?

AntiTcb commented:
Sounds good to me.

pegasy committed:
removed Guild from class name for ChannelCategory
Renamed all properties to use Category instead of Parent
Throw exception on GetUsers / GetInvites etc for categories

MinnDevelopment commented:
Wouldn't it be more consistent to call it CategoryChannel then if you have TextChannel and VoiceChannel?

pegasy commented:
CategoryChannel makes sense to call it from the guild (CreateCategoryChannel) but looks really odd to me to have channel.CategoryChannel but might just be me.
Who decides on naming convention?

FiniteReality commented:
`channel.Category` makes sense. `SocketCategoryChannel`, `RestCategoryChannel` and `ICategoryChannel` also make sense.

pegasy committed:
Renamed classes / properties / methods to use CategoryChannel instead of ChannelCategory to be consistant with how text / voice channels are named.

LDSpits commented:
@foxbot Is it alright if i finish this pull request? Since it looks like it has gone stale.

FiniteReality commented:
I'm fine with somebody else continuing it, but it's probably best if we wait a bit longer for @pegasy to respond first. It looks like they've committed recently (within the past ~5 days) so they may have simply missed us.

WittyAdrian commented:
@FiniteReality I think you misread, as far as I can see pegasy last committed over a month ago.

khionu commented:
To this PR, yes. To GitHub at all, no. The thread has been bumped, so the time to wait has been reset.

DeltaDizzy commented on src/Discord.Net.Core/Entities/Guilds/IGuild.cs:
```diff
@@ -93,6 +94,8 @@ public interface IGuild : IDeletable, ISnowflakeEntity
         /// <summary> Creates a new voice channel. </summary>
         Task<IVoiceChannel> CreateVoiceChannelAsync(string name, RequestOptions options = null);
+        /// <summary> Creates a new channel category. </summary>
+        Task<ICategoryChannel> CreateCategoryChannelAsync(string name, RequestOptions options = null);
```
its talking about creating a channel within a category.

FiniteReality commented:
Ok. I'm fairly confident that this PR is stale now. If anybody wants to pick it up, feel free.
